tgui/packages/tgui/interfaces/MCDependencyDebug.tsx

```diff
@@ -0,0 +1,303 @@
+};
+
+type SubsystemMap = {
+  [subsystemName: string]: Subsystem;
+};
+
+function EvaluateSubsystemLayer(
+  subsystem: Subsystem,
+  depth: number,
+  data: SubsystemLayer,
+) {
+  data[subsystem.name] = Math.max(depth, data[subsystem.name] || 0);
+  for (let i = 0; i < subsystem.dependents.length; i++) {
+    EvaluateSubsystemLayer(subsystem.dependents[i], depth + 1, data);
+  }
+}
+
+export const MCDependencyDebug = (props, context) => {
+  const { data } = useBackend<DependencyData>();
+  const { subsystems } = data;
+  const connectionDom = useRef<ConnectionRef>({});
+  const [connectionData, setConnectionData] = useState<ConnectionData>({});
+
+  useLayoutEffect(() => {
+    const doms = connectionDom.current;
+    const newConnectionData: ConnectionData = {};
+    for (let i = 0; i < subsystems.length; i++) {
+      const subsystem = subsystems[i];
+      const domElements = doms[subsystem.name];
+      if (domElements.input === undefined || domElements.output === undefined) {
+        continue;
+      }
+      newConnectionData[subsystem.name] = {
+        input: getPosition(domElements.input),
+        output: getPosition(domElements.output),
+      };
+    }
+    setConnectionData(newConnectionData);
+  }, [subsystems]);
+
+  const subsystemsLayered = useMemo(() => {
+    const namesToSubsystem: SubsystemMap = {};
+    const subsystemsGraph: Subsystem[] = [];
+
+    for (let i = 0; i < subsystems.length; i++) {
+      const subsystem = subsystems[i];
+      const subsystemNode = {
+        name: subsystem.name,
+        dependents: [],
+      };
+      namesToSubsystem[subsystem.name] = subsystemNode;
+      subsystemsGraph.push(subsystemNode);
+    }
+
+    for (let i = 0; i < subsystems.length; i++) {
+      const subsystem = subsystems[i];
+      subsystemsGraph[i].dependents = subsystem.dependents.map(
+        (val) => namesToSubsystem[val],
+      );
+    }
+
+    const subsystemLayer: SubsystemLayer = {};
+    for (let i = 0; i < subsystemsGraph.length; i++) {
+      const subsystem = subsystemsGraph[i];
+      EvaluateSubsystemLayer(subsystem, 1, subsystemLayer);
+    }
+    return subsystemLayer;
+  }, [subsystems]);
+
+  const connections: Connection[] = [];
+  for (let i = 0; i < subsystems.length; i++) {
+    const subsystem = subsystems[i];
+    const from = connectionData[subsystem.name];
+    if (from === undefined) {
+      continue;
+    }
+    for (let j = 0; j < subsystem.dependents.length; j++) {
+      const outputSubsystem = subsystem.dependents[j];
+      const target = connectionData[outputSubsystem];
+      if (target === undefined) {
+        continue;
+      }
+
+      connections.push({
+        from: from.output,
+        to: target.input,
+        style: ConnectionStyle.SUBWAY_SHARP,
+        index: i,
+        color: `hsl(${60 + 5 * (i % 30)}, 50%, ${50 + (i % 30)}%)`,
+      });
+    }
+  }
+
+  return (
+    <Window width={1200} height={800} title={'Subsystem Dependency Graph'}>
```

Comment:
Not a problem, just curious: Is `{'` is an IDE thing? Could just be double quotes